pcap-dpdk.c

```diff
@@ -0,0 +1,742 @@
+	uint64_t delta_usec = delta_tm.tv_sec*1e6+delta_tm.tv_usec;
+	uint64_t delta_bit = (pd->curr_stats.ibytes-pd->prev_stats.ibytes)*8;
+	RTE_LOG(INFO, USER1, "delta_usec: %-10"PRIu64" delta_pkt: %-10"PRIu64" delta_bit: %-10"PRIu64"\n", delta_usec, delta_pkt, delta_bit);
+	pd->pps = (uint64_t)(delta_pkt*1e6f/delta_usec);
+	pd->bps = (uint64_t)(delta_bit*1e6f/delta_usec);
+	nic_stats_display(pd);
+	pd->prev_stats = pd->curr_stats;
+	pd->prev_ts = pd->curr_ts;
+	return 0;
+}
+
+static int pcap_dpdk_setnonblock(pcap_t *p, int fd _U_){
+	pcap_fmt_errmsg_for_errno(p->errbuf, PCAP_ERRBUF_SIZE,
+		errno, "dpdk error: setnonblock not support");
+	return 0;
+}
+
+static int pcap_dpdk_getnonblock(pcap_t *p){
+	pcap_fmt_errmsg_for_errno(p->errbuf, PCAP_ERRBUF_SIZE,
+		errno, "dpdk error: getnonblock not support");
+	return 0;
+}
+
+static int check_link_status(uint16_t portid, struct rte_eth_link *plink)
+{
+	uint8_t count = 0;
+	int is_port_up = 0;
+	int max_check_time = 50;
+	int check_interval = 100; // 100ms
+	for (count = 0; count <= max_check_time; count++) {
+		memset(plink, 0, sizeof(struct rte_eth_link));
+		rte_eth_link_get_nowait(portid, plink);
+		if (plink->link_status == ETH_LINK_UP)
+		{
+			is_port_up = 1;
+			break;
+		}else{
+			rte_delay_ms(check_interval);
+		}
+	}
+	return is_port_up;
+}
+
+// return portid by device name, otherwise return -1
+static uint16_t portid_by_device(char * device)
+{
+	uint16_t ret = -1;
+	int len = strlen(device);
+	int prefix_len = strlen(DPDK_PREFIX);
+	unsigned long ret_ul = 0L;
+
+	if (len<=prefix_len || strncmp(device, DPDK_PREFIX, prefix_len)) // check prefix dpdk:
+	{
+		return ret;
+	}
+	if (device[prefix_len]>='0' && device[prefix_len]<='9')
+	{ // is digital
+		ret_ul = strtoul(&(device[prefix_len]), NULL, 10);
+		ret = (uint16_t)ret_ul;
+	}
+	return ret;
+}
+
+static int parse_dpdk_cfg(char* dpdk_cfg,char** dargv)
+{
+	int cnt=0;
+	memset(dargv,0,sizeof(dargv[0])*DPDK_ARGC_MAX);	
+	//current process name
+	int skip_space = 1;
+	int i=0;
+	RTE_LOG(INFO, USER1,"dpdk cfg: %s\n",dpdk_cfg);
+	// find first non space char
+	// The last opt is NULL
+	for (i=0;dpdk_cfg[i] && cnt<DPDK_ARGC_MAX-1;i++){
+		if (skip_space && dpdk_cfg[i]!=0x20){ // not space
+			skip_space=!skip_space; // skip normal char
+			dargv[cnt++] = dpdk_cfg+i;
+		}
+		if (!skip_space && dpdk_cfg[i]==0x20){ // fint a space
+			dpdk_cfg[i]=0x00; // end of this opt
+			skip_space=!skip_space; // skip space char
+		}
+	}
+	dargv[cnt]=NULL;
+	return cnt;
+}
+static int pcap_dpdk_activate(pcap_t *p)
+{
+	struct pcap_dpdk *pd = p->priv;
+	pd->orig = p;
+	int ret = PCAP_ERROR;
+	uint16_t nb_ports=0;
+	uint16_t portid=-1;
+	unsigned nb_mbufs = DPDK_NB_MBUFS;
+	struct rte_eth_rxconf rxq_conf;
+	struct rte_eth_txconf txq_conf;
+	struct rte_eth_conf local_port_conf = port_conf;
+	struct rte_eth_dev_info dev_info;
+	int is_port_up = 0;
+	struct rte_eth_link link;
+	if (p == NULL)
+	{
+		return PCAP_ERROR;
```

Comment:
1. This should never happen - the activate routine is only called by `pcap_activate()`, and, before it calls it, it'll call `pcap_check_activated()`, which will check `p->activated`, and that will crash, on most if not all platforms, if `p` is null.

1. If you return `PCAP_ERROR`, you *have* to set `p->errbuf` with an error message.

Reply:
OK, the check if (p==NULL) will be removed safely.